arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,749 @@
+        break;
+      }
+      case AV_PIX_FMT_MONOWHITE:
+      case AV_PIX_FMT_MONOBLACK:
+      {
+        num_pixels = 1;
+        pixel_trait = vital::image_pixel_traits_of<bool>();
+        direct_copy = true;
+        break;
+      }
+    }
+    if (direct_copy)
+    {
+      int size = avpicture_get_size(enc->pix_fmt, width, height);
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+
+      AVPicture frame;
+      avpicture_fill(&frame, (uint8_t*)d->current_image_memory->data(), enc->pix_fmt, width, height);
+      av_picture_copy(&frame, (AVPicture*)d->f_frame, enc->pix_fmt, width, height);
+    }
+    else
+    {
+      int size = width * height * num_pixels;
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+
+      d->f_software_context = sws_getCachedContext(
+        d->f_software_context,
+        width, height, enc->pix_fmt,
+        width, height, AV_PIX_FMT_RGB24,
+        SWS_BILINEAR,
+        NULL, NULL, NULL);
+
+      if (!d->f_software_context)
+      {
+        LOG_ERROR(this->logger(), "Couldn't create conversion context");
+        return nullptr;
+      }
+
+      AVPicture rgb_frame;
+      avpicture_fill(&rgb_frame, (uint8_t*)d->current_image_memory->data(), AV_PIX_FMT_RGB24, width, height);
+
+      sws_scale(d->f_software_context,
+        d->f_frame->data, d->f_frame->linesize,
+        0, height,
+        rgb_frame.data, rgb_frame.linesize);
+    }
+
+    vital::image image(
+      d->current_image_memory,
+      d->current_image_memory->data(),
+      width, height, num_pixels,
+      num_pixels, num_pixels * width, num_pixels * width*height
+    );
+    d->current_image = std::make_shared<vital::simple_image_container>(vital::simple_image_container(image));
+  }
+
+  return d->current_image;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::timestamp
+ffmpeg_video_input
+::frame_timestamp() const
+{
+  if (!this->good())
+  {
+    return{};
+  }
+
+  // We don't always have all components of a timestamp, so start with
+  // an invalid TS and add the data we have.
+  kwiver::vital::timestamp ts;
+  ts.set_frame(d->frame_number());
+
+  return ts;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::metadata_vector
+ffmpeg_video_input
+::frame_metadata()
+{
+  LOG_INFO(this->logger(), "Metadata access isn't supported yet");
+  return kwiver::vital::metadata_vector();
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::metadata_map_sptr
+ffmpeg_video_input
+::metadata_map()
+{
+  LOG_INFO(this->logger(), "Metadata access isn't supported yet");
+  return std::make_shared<kwiver::vital::simple_metadata_map>(
+    kwiver::vital::simple_metadata_map());
+}
+
+
+// ------------------------------------------------------------------
+bool
+ffmpeg_video_input
+::end_of_video() const
+{
+  return false;
```

Comment:
Is there a way to tell from the FFmpeg API if we are at the end of the video? It looks like the old VXL implementation set a flag to mark it at the end if the decoding failed but I am not sure if that is the correct approach.

Reply:
Please take a look at the VXL implementation and reproduce here since the `end_of_video()` method is used in multiple places in both maptk and kwiver.